tendermint/src/rpc/event_listener.rs

```diff
@@ -0,0 +1,241 @@
+    ///Subscribe to all blocks
+    BlockSubscription,
+}
+
+impl EventSubscription {
+    ///Convert the query enum to a string
+    pub fn as_str(&self) -> &str {
+        match self {
+            EventSubscription::TransactionSubscription => "tm.event='Tx'",
+            EventSubscription::BlockSubscription => "tm.event='NewBlock'",
+        }
+    }
+}
+
+/// Event Listener over websocket.
+/// See: <https://docs.tendermint.com/master/rpc/#/Websocket/subscribe>
+pub struct EventListener {
+    socket: async_tungstenite::WebSocketStream<TokioAdapter<TcpStream>>,
+}
+
+impl EventListener {
+    /// Constructor for event listener
+    pub async fn connect(address: net::Address) -> Result<EventListener, RPCError> {
+        let (host, port) = match address {
+            net::Address::Tcp { host, port, .. } => (host, port),
+            other => {
+                return Err(RPCError::invalid_params(&format!(
+                    "invalid RPC address: {:?}",
+                    other
+                )));
+            }
+        };
+        //TODO This doesn't have any way to handle a connection over TLS
+        let (ws_stream, _unused_tls_stream) =
+            connect_async(&format!("ws://{}:{}/websocket", host, port)).await?;
+        Ok(EventListener { socket: ws_stream })
+    }
+    //TODO Have a query type instead of a string
+    /// Subscribe to event query stream over the websocket
+    pub async fn subscribe(&mut self, query: EventSubscription) -> Result<(), RPCError> {
+        self.socket
+            .send(Message::text(
+                subscribe::Request::new(query.as_str().to_owned()).into_json(),
+            ))
+            .await?;
+        Ok(())
+    }
+
+    /// Get the next event from the websocket
+    pub async fn get_event(&mut self) -> Result<Event, RPCError> {
+        let msg = self
+            .socket
+            .next()
+            .await
+            .ok_or_else(|| RPCError::websocket_error("web socket closed"))??;
+        match serde_json::from_str::<JsonRPCBlockResult>(&msg.to_string()) {
+            Ok(data) => {
+                let block_result = data.into_result()?;
+                Ok(Event::JsonRPCBlockResult(Box::new(block_result)))
+            }
+            Err(e) => {
+                dbg!("Error:");
+                dbg!(e);
+                dbg!("msg:");
+                dbg!(&msg.to_string());
+                match serde_json::from_str::<JsonRPCTransactionResult>(&msg.to_string()) {
+                    Ok(data) => {
+                        let tx_result = data.into_result()?;
+                        Ok(Event::JsonRPCTransactionResult(Box::new(tx_result)))
+                    }
+                    Err(_) => match serde_json::from_str::<serde_json::Value>(&msg.to_string()) {
+                        Ok(json_value) => Ok(Event::GenericJSONEvent(json_value)),
+                        Err(_) => Ok(Event::GenericStringEvent(msg.to_string())),
+                    },
+                }
```

Comment:
I think ideally, we would not match case a lot here but instead of jsonrpc wrapping `JsonRPCBlockResult`, `JsonRPCTransactionResult` separately, we should wrap the whole `Event` enum and make it `impl response::Response`. serde can very well deal with [enums](https://serde.rs/enum-representations.html) so there is no need to try each variant separately. No need to block this PR on this though.